### ISSUE TITLE
Feature Request: Check for version-specific Bash features

### ISSUE THREAD

**Issue by gwerbin** [koalaman/shellcheck#2850]
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

GNU Bash has made some significant changes across versions. Sometimes we want to write scripts that work on a wide range of versions, e.g. MacOS 12.4 still ships with Bash 3.x.

There are several feature matrices out there for Bash, but they require manual verification and inspection, which is exactly what Shellcheck is good for!

Documentation resources include:

* The official NEWS document: https://git.savannah.gnu.org/cgit/bash.git/tree/NEWS
* The official COMPAT document: https://git.savannah.gnu.org/cgit/bash.git/tree/COMPAT
* Other 3rd-party documents:
  * http://mywiki.wooledge.org/BashFAQ/061
  * https://tldp.org/LDP/abs/html/bash2.html (does not include Bash 5)

It would be very useful if Shellcheck allowed us to specify a specific Bash _version_, e.g. `shell=bash:3.x`, and could then check for feature compatibility.

For example, if I declare `shell=bash:3.`, the `mapfile` and `readarray` builtins should be prohibited unless I define my own shell functions of the same name. If the user happens to have an executable program on their system of the same name, they could opt to disable this particular incompatibility warning (or they could use `env mapfile` to work around the ambiguity).

Of course this could be extended to other shells in principle, but my immediate use case is Bash, and I assume that probably applies to most other Shellcheck users out there.

Finally, thank you for the excellent tool! Shellcheck has been a productivity booster and reliable safeguard in my work, and I am grateful for its existence.

**Comment by dalisoft** [comment]
Could be nice to see such feature

**Comment by ithinkihaveacat** [comment]
Maybe a nice way to implement this is to check if the script is already checking against `$BASH_VERSINFO`. If it is, then `shellcheck` enables feature version checking. That is, if a script starts with:

```sh
if ((BASH_VERSINFO[0] < 3)); then
  echo "$(basename "$0"): requires bash 4.0 or higher (found ${BASH_VERSION})" >&2
  exit 1
fi
```

… then shellcheck verifies that the remainder of the script is compatible with bash 4.0 or higher. (Adjust as appropriate if the check triggers conditional code, etc.)

This has the advantage of not requiring any new syntax. It also handles a related problem (bug?): scripts using features added in a very recent version of bash should probably guard this with a version check. (If my script inadvertently uses a bash 5.3 feature, in most cases I would like to be warned by shellcheck if there's no guard around it.)
